crates/fuel-core/src/service/adapters/graphql_api/off_chain.rs

```diff
@@ -187,6 +202,72 @@ impl OffChainDatabase for OffChainIterableKeyValueView {
     fn message_is_spent(&self, nonce: &Nonce) -> StorageResult<bool> {
         self.message_is_spent(nonce)
     }
+
+    fn balance(
+        &self,
+        owner: &Address,
+        asset_id: &AssetId,
+        base_asset_id: &AssetId,
+    ) -> StorageResult<TotalBalanceAmount> {
+        let coins = self
+            .storage_as_ref::<CoinBalances>()
+            .get(&BalancesKey::new(owner, asset_id))?
+            .unwrap_or_default()
+            .into_owned() as TotalBalanceAmount;
+
+        if base_asset_id == asset_id {
+            let messages = self
+                .storage_as_ref::<MessageBalances>()
+                .get(owner)?
+                .unwrap_or_default()
+                .into_owned() as TotalBalanceAmount;
+
+            let total = coins.checked_add(messages).ok_or(anyhow::anyhow!(
+                "Total balance overflow: coins: {coins}, messages: {messages}"
+            ))?;
+
+            debug!(%coins, %messages, total, "total balance");
+            Ok(total)
+        } else {
+            debug!(%coins, "total balance");
+            Ok(coins)
+        }
+    }
+
+    fn balances(
+        &self,
+        owner: &Address,
+        base_asset_id: &AssetId,
+    ) -> StorageResult<BTreeMap<AssetId, TotalBalanceAmount>> {
+        let mut balances = BTreeMap::new();
```

Comment:
It would require some shenanigans to lazily "inject" the balance of messages into the coin with base asset id. Probably we'd need to "collect" the coins one way or another in the process.

I'm currently verifying another idea.

Reply:
To summarize, there are at least 3 ways to go:
1. leave it as is - collect in memory, adjust for base_asset_id
2. let the indexation module handle updates of coins with base asset id ([implemented on separate branch](https://github.com/FuelLabs/fuel-core/compare/1965_balances_cache...1965_balances_cache_include_messages?expand=1), since I'm not sure if this is a good design)
3. lean towards [your suggestion](https://github.com/FuelLabs/fuel-core/pull/2383#discussion_r1823128502) about single DB, which I think I like even less than pt. 2 above :)

Reply:
4th approach has been implemented :)
1. Get the non_retryable value of a message
2. Get the value of coin with base_asset_id
3. Make an iterator that yields sum of the above
4. Get iterator for coins with asset id different that base asset id
5. Chain iterators from pt. 4 and 5, respecting the requested direction